lib/lists.nix

```diff
@@ -95,25 +95,84 @@ rec {
+  /* Same as `imap0`, but indices start from 1. O(n) complexity.
 
      Type: imap1 :: (int -> a -> b) -> [a] -> [b]
 
      Example:
        imap1 (i: v: "${v}-${toString i}") ["a" "b"]
        => [ "a-1" "b-2" ]
   */
-  imap1 = f: list: genList (n: f (n + 1) (elemAt list n)) (length list);
+  imap1 = f: list:
+    let eAt = elemAt list; in genList (n: f (n + 1) (eAt n)) (length list);
+
+  /* Like `filter`, but with an index. O(n) complexity.
+
+     Note that `ifilter0` does not stack allocate any intermediate lists.
+
+     Type: filter :: (int -> a -> bool) -> [a] -> [a]
+
+     Example:
+       ifilter0 (i: v: i == 0 || v > 2) [ 1 2 3 ]
+       => [ 1 3 ]
+  */
+  ifilter0 =
+    ipred:
+    list:
+    let listLength = length list; in
+    if listLength <= 0 then list else
+    /* For a function `generator :: n -> a` and variable `count :: n`,
+       list `list :: [a]` of `list = genList generator count` memoizes
+       `generator` over the domain ((domain of `list`) union `range 0 count`).
+       Function `(elemAt list) :: n -> a` provides the same interface as
+       `generator`.
+    */
+    let
+      # View `list` as a memoization.
+      # `nToE` maps `n` (a `list` index) to `e` (a `list` element). Memoized.
+      nToE = elemAt list;
+      # `nToKeep` maps `n` (a `list` index) to `keep` (whether a `list` element
+      # `e` should be kept, according to `ipred`). Memoized.
+      nToKeep = elemAt nToKeepList;
+      nToKeepList = genList (n: ipred n (nToE n)) listLength;
+      # `keptNToEKept` maps `keptN` (a `keptList` index) to `eKept` (a kept
+      # `list` element). Our final result is this memoization, viewed as a list.
+      keptList = keptNToEKeptList;
+      # The length of `keptList` will be the count of kept elements in `list`.
+      keptLength =
+        foldl' (count: keep: count + (if keep then 1 else 0)) 0 nToKeepList;
+      # `nToNKept` maps `n` (a `list` index) to `nKept` (the closest `list`
+      # index greater than or equal to `n` for which `keep` is true).
+      nToNKept = n: if nToKeep n then n else nToNKept (n + 1);
+      # `keptNToNUniq` maps `keptN` (a `keptList` index) to `nKept` (a `list`
+      # index with a kept element). Memoized.
+      keptNToNKept = elemAt keptNToNKeptList;
+      keptNToNKeptList = let keptNToNKeptUnmemoized = keptN:
+        # Base case. Begin searching for a kept element from the beginning.
+        if keptN == 0 then nToNKept 0 else let
+          # To get `keptN`'s `nKept`, we need the previous `keptN`'s `nKept`.
+          prevKeptN = keptN - 1;
+          # This recursion is cheap thanks to memoization.
+          prevNKept = keptNToNKept prevKeptN;
+          # Find the next `nKept`, not `prevNKept` again.
+          n = prevNKept + 1;
+        in nToNKept n
+      ; in genList keptNToNKeptUnmemoized keptLength;
+      keptNToEKeptList = let keptNToEKeptUnmemoized = keptN:
+        nToE (keptNToNKept keptN)
+      ; in genList keptNToEKeptUnmemoized keptLength;
```

Comment:
I'm not sure if these variable names and comments make it easier or harder to understand!

How about renaming
- `nToKeepList` -> `keepOrNot`
- `keptLength` -> `resultLength`
- `nToNKept` -> `nextResultIndex`
- `keptNToNKeptList` -> `resultIndices`
- `keptNToEKeptList` -> `result`

Also, how about adding some graphics:
```
Example: Filtering for only uppercase letters in [ "b" "A" "f" "B" "E" "c" ]

       unfiltered list indices: 0 1 2 3 4 5
      unfiltered list elements: b A f B E c
       include condition holds: - Y - Y Y -  -> result length is 3
                                               
         filtered list indices:   0   1 2
 unfiltered indices to include:   1   3 4
                                  ^   ^ ^
    traverse condition list to find   Continue traversing condition list after the
the first Y, return the index of it   previously included index to find the next Y
                                 
        filtered list elements:   A   B E
```

Reply:
My idea with the naming was to keep it clear what type of data you were dealing with by using consistent "conversion" functions, especially when different index styles are being processed.

Visuals in the documentation is a good idea.